File: Plugins/AudioPlayer/src/android/app/src/main/java/com/shortsands/audioplayer/AudioTOCTestament.java
package com.shortsands.audioplayer;

/**
 * Created by garygriswold on 8/30/17.
 */
import android.database.Cursor;
import android.util.Log;
import com.shortsands.utility.Sqlite3;
import java.util.HashMap;
import java.util.Iterator;

class AudioTOCTestament {

    private static String TAG = "AudioTOCTestament";

    AudioTOCBible bible;
    String damId;
    String dbpLanguageCode;
    String dbpVersionCode;
    private String collectionCode;
    private String mediaType;
    HashMap<String, AudioTOCBook> booksById;
    private HashMap<Integer, AudioTOCBook> booksBySeq;

    AudioTOCTestament(AudioTOCBible bible, Sqlite3 database, Cursor cursor) {
        this.bible = bible;
        this.booksById = new HashMap<String, AudioTOCBook>();
        this.booksBySeq = new HashMap<Integer, AudioTOCBook>();
        this.damId = cursor.getString(0);
        this.collectionCode = cursor.getString(1);
        this.mediaType = cursor.getString(2);
        this.dbpLanguageCode = cursor.getString(3);
        this.dbpVersionCode = cursor.getString(4);

        String query = "SELECT bookId, bookOrder, bookName, numberOfChapters" +
            " FROM AudioBook" +
            " WHERE damId = ?" +
            " ORDER BY bookOrder";
        Object[] values = new Object[1];
        values[0] = this.damId;
        try {
            Cursor cursor2 = database.queryV0(query, values);
            int rowNum = 0;
            while(cursor2.moveToNext()) {
                AudioTOCBook book = new AudioTOCBook(this, rowNum, cursor2);
                this.booksById.put(book.bookId, book);
                this.booksBySeq.put(book.sequence, book);
                rowNum++;
            }
        } catch (Exception err) {
            Log.d(TAG,"ERROR " + Sqlite3.errorDescription(err));
        }
    }

    //deinit {
    //    print("***** Deinit TOCAudioTOCBible *****")
    //}

    AudioReference nextChapter(AudioReference ref) {
        AudioTOCBook book = ref.tocAudioBook;
        if (ref.chapterNum() < book.numberOfChapters) {
            int next = ref.chapterNum() + 1;
            return AudioReference.factory(ref.tocAudioBook, next, ref.fileType);
        } else {
            AudioTOCBook nextBook = this.booksBySeq.get(ref.sequenceNum() + 1);
            if (nextBook != null) {
                return AudioReference.factory(nextBook, 1, ref.fileType);
            }
        }
        return null;
    }

    AudioReference priorChapter(AudioReference ref) {
        int prior = ref.chapterNum() - 1;
        if (prior > 0) {
            return AudioReference.factory(ref.tocAudioBook, prior, ref.fileType);
        } else {
            AudioTOCBook priorBook = this.booksBySeq.get(ref.sequenceNum() - 1);
            if (priorBook != null) {
                return AudioReference.factory(priorBook, priorBook.numberOfChapters, ref.fileType);
            }
        }
        return null;
    }

    String getBookList() {
        StringBuffer array = new StringBuffer();
        Iterator<AudioTOCBook> it = this.booksBySeq.values().iterator();
        while(it.hasNext()) {
            array.append(",");
            array.append(it.next().bookId);
        }
        return array.toString();
    }

    public String toString() {
        String str = "damId=" + this.damId +
                "\n languageCode=" + this.dbpLanguageCode +
                "\n versionCode=" + this.dbpVersionCode +
                "\n mediaType=" + this.mediaType +
                "\n collectionCode=" + this.collectionCode;
        return str;
    }
}

File: Plugins/AudioPlayer/src/ios/AudioPlayer/AudioTOCTestament.swift
//
//  AudioTOCTestament.swift
//  AudioPlayer
//
//  Created by Gary Griswold on 8/8/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//
#if USE_FRAMEWORK
import Utility
#endif

class AudioTOCTestament {
    
    let bible: AudioTOCBible
    let damId: String
    let dbpLanguageCode: String
    let dbpVersionCode: String
    let collectionCode: String
    let mediaType: String
    var booksById: Dictionary<String, AudioTOCBook>
    var booksBySeq: Dictionary<Int, AudioTOCBook>

    init(bible: AudioTOCBible, database: Sqlite3, dbRow: [String?]) {
        self.bible = bible
        self.booksById = Dictionary<String, AudioTOCBook>()
        self.booksBySeq = Dictionary<Int, AudioTOCBook>()
        self.damId = dbRow[0]!
        self.collectionCode = dbRow[1]!
        self.mediaType = dbRow[2]!
        self.dbpLanguageCode = dbRow[3]!
        self.dbpVersionCode = dbRow[4]!

        let query = "SELECT bookId, bookOrder, numberOfChapters" +
            " FROM AudioBook" +
            " WHERE damId = ?" +
            " ORDER BY bookOrder"
        do {
            let resultSet = try database.queryV1(sql: query, values: [self.damId])
            for row in resultSet {
                let book = AudioTOCBook(testament: self, dbRow: row)
                self.booksById[book.bookId] = book
                self.booksBySeq[book.sequence] = book
            }
        } catch let err {
            print("ERROR \(Sqlite3.errorDescription(error: err))")
        }
    }
    
    deinit {
       print("***** Deinit TOCAudioTOCBible *****") 
    }
    
    func nextChapter(reference: AudioReference) -> AudioReference? {
        let ref = reference
        let book = ref.tocAudioBook
        if (ref.chapterNum < book.numberOfChapters) {
            let next = ref.chapterNum + 1
            return AudioReference(book: ref.tocAudioBook, chapterNum: next, fileType: ref.fileType)
        } else {
            if let nextBook = self.booksBySeq[reference.sequenceNum + 1] {
                return AudioReference(book: nextBook, chapter: "001", fileType: ref.fileType)
            }
        }
        return nil
    }
    
    func priorChapter(reference: AudioReference) -> AudioReference? {
        let ref = reference
        let prior = ref.chapterNum - 1
        if (prior > 0) {
            return AudioReference(book: ref.tocAudioBook, chapterNum: prior, fileType: ref.fileType)
        } else {
            if let priorBook = self.booksBySeq[reference.sequenceNum - 1] {
                return AudioReference(book: priorBook,
                                 chapterNum: priorBook.numberOfChapters, fileType: ref.fileType)
            }
        }
        return nil
    }
    
    func getBookList() -> String {
        var array = [String]()
        for (_, book) in self.booksBySeq {
            array.append(book.bookId)
        }
        return array.joined(separator: ",")
    }
    
    func toString() -> String {
        let str = "damId=" + self.damId +
                "\n languageCode=" + self.dbpLanguageCode +
                "\n versionCode=" + self.dbpVersionCode +
                "\n mediaType=" + self.mediaType +
                "\n collectionCode=" + self.collectionCode
        return str
    }
}
